Align macros usage with Vanilla standards

britneywwc opened the pull request:
## Done
- Use Vanilla defined macros for hero section

## To-do
- [x] Figure out a better way to import macros from Vanilla, currently just copy-pasting the macros code directly
- [x] Remove inline styles and use utilities where possible (done for modal styles)
- [x] Update highlights macro to limit list items (refer to [Tiered List](https://github.com/canonical/vanilla-framework/blob/3885de3af2c1393b5a51302142b3a17dc6e8dd4d/templates/_macros/tiered-list.jinja#L59))
- [x] Update highlights list parent and children tags
- [x] Refactor text macros to accept paragraphs as slots
- [x] Refactor image macro to use the standard image loader
- [x] Align with quotes section that is currently WIP: https://github.com/canonical/vanilla-framework/pull/5275

## QA

- Go to https://ubuntu-com-14166.demos.haus/case-study/example
- [Figma design](https://www.figma.com/design/AeIJ3GsqnJCHtiBYhkTuTV/24.10-Case-study-template?node-id=1-716&t=43Yuf8LEb6wYh1pC-0)

## Issue / Card

Fixes [WD-14166](https://github.com/canonical/ubuntu.com/pull/14166)

## Screenshots

[If relevant, please include a screenshot.]


## Help

[QA steps](https://discourse.canonical.com/t/qa-steps/152) - [Commit guidelines](https://discourse.canonical.com/t/commit-guidelines/148)


[WD-14166]: https://warthogs.atlassian.net/browse/WD-14166?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://ubuntu-com-14166.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/10020)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

britneywwc commented:
Hey @jmuzina @pastelcyborg! I will be addressing comments of the macros usage in this PR. 
Is there a suggested way to import macros? The Vanilla build does not include macro files thus I'm not able to import them through npm, I'm currently just copy-pasting the macro codes directly.

jmuzina commented on templates/case-studies/index.html:
```diff
@@ -4,15 +4,27 @@
-  <!-- Hero section-->
+  
+  <!-- Hero section-->  
   {% import "case-studies/templates/_macros-hero.jinja" as hero_template %}
```
We have been importing our macros at the top of the templates that use them, like this:

https://github.com/canonical/vanilla-framework/blob/main/templates/docs/examples/patterns/hero/hero-50-50.html

britneywwc commented:
Thanks for all the suggestions! I have addressed all the comments in this PR and the feature branch, I've also consolidated them in the list of "To-do" under the PR description.

As for the quotes, I will wait for the Vanilla PR to be merged since the quotes that are being used in the case studies are the same.

jmuzina commented:
Heads up @britneywwc : https://github.com/canonical/vanilla-framework/pull/5275 has macro markup that is ready for use, but there are a couple issues related to HR spacing that are blocking it - they will be solved by https://github.com/canonical/vanilla-framework/pull/5241.

jmuzina commented on templates/case-studies/index.html:
```diff
@@ -1,72 +1,100 @@
-    )
-  }}
+  {%- call(slot) text(title="You can’t explore space without infrastructure.") -%}
+    {%- if slot == 'paragraph_1' -%}
```
Following up on our discussion in MM, I don't think these paragraphs really need to be separate slots in a range. Maybe you can make a single slot for paragraph content, and pass all of these in there?

My reasoning is that the paragraph-handling part of the text macro seems to be simple enough that the extra work of calling a macro isn't really worth it for these paragraphs, since the macro isn't adding or abstracting all that much.

britneywwc replied:
Thanks for the suggestion, I definitely agree that it is better to pass it in a single slot. The only issue is that ideally we'd like the content writers to use these macros, so if they were to pass the text chunks in a single slot they would have to manually define breakpoints/paragraphs (e.g using `br` or `\n`). Thus the separate slots were introduced to deal with paragraphing each text chunk.

Is there a better approach to do so if we want to eliminate the need of content writers to define breakpoints?

jmuzina replied:
> Is there a better approach to do so if we want to eliminate the need of content writers to define breakpoints?

Like we talked about in MM, while content writers should define breakpoints (with line breaks in the paragraphs they type), I don't see them doing it in HTML. 

I don't currently see a better solution for this, but if we absolutely had to support something like that we could maybe find a way to make Flask read in the page breaks and add a new paragraph for each one

jmuzina commented on templates/case-studies/index.html:
```diff
@@ -1,72 +1,100 @@
+      <p>Architecture, design, and deployment of Canonical’s distributions of Ceph and Kubernetes.</p>
+    {%- endif -%}
+    {%- if slot == 'highlights_item_content_2' -%}
+      <p>Timely updates and 24/7 monitoring of missioncritical equipment.</p>
```
Missing a space between "mission" and "critical"

```suggestion
      <p>Timely updates and 24/7 monitoring of mission critical equipment.</p>
```

jmuzina commented on templates/macros/_macros-image.jinja:
```diff
@@ -0,0 +1,22 @@
+  <section class="p-strip">
+    <div class="row">
+      {% if has_image %}
+        <div class="col u-align--center u-vertically-center u-no-padding--bottom p-image-container--cinematic">
```
`p-image-container` requires the image inside to have the class `p-image-container__image`, but at the level of abstraction in this macro, the macro has no control over the class on the image that is passed in.

Probably the caller should be responsible for passing in `p-image-container` or something else

[Docs](https://vanillaframework.io/docs/patterns/images#image-container-with-aspect-ratio)

britneywwc replied:
I'll add a note in `_macros-image.jinja` to state that the caller must pass in `p-image-container` for now. I think it would be best practice to set the class of the image being passed if possible, but I'm not sure how to enforce that within this macro. If you have any suggestions please let me know as well

britneywwc commented on templates/macros/_macros-highlights.jinja:
```diff
@@ -0,0 +1,33 @@
+      <p class="p-text--small-caps">Highlights</p>
+      <hr />
+    </div>
+    {% if num_items == 3 %}
```
We are restricting to 3/4 highlight items only, I can add an `else` catch just in case. @juanruitina could you weigh in on how we'd want to handle the columns for unsupported user values?

juanruitina replied:
If there were more than 4 items, then it could turn into a regular list. But I don't know if we want to allow that (I don't have strong opinions against it. I defer to @lyubomir-popov on this.

lyubomir-popov replied:
I think that's an elegant fallback

britneywwc commented on templates/macros/_macros-text.jinja:
```diff
@@ -0,0 +1,26 @@
+          {%- set has_paragraph_content = paragraph_content|trim|length > 0 %}
+
+          {%- if has_paragraph_content -%}
+            <div>{{ paragraph_content }}</div>
```
I added the `div` element for spacing purposes, with just `p` there is less spacing between paragraphs. Maybe I can wrap it in `div` and `p` in the macros?

britneywwc replied:
I added the `div` initially for spacing purposes between paragraphs. I think based on @jmuzina's [suggestions](https://github.com/canonical/ubuntu.com/pull/14166#discussion_r1717456885), I will change the implementation of paragraphs and just wrap the whole thing in a `p`.

jmuzina commented on templates/case-studies/example.html:
```diff
@@ -1,72 +1,85 @@
     {%- if slot == 'image_content' %}
       {{ image(url="https://assets.ubuntu.com/v1/c3f56868-Global_Mars_in_colour 1.png",
-                alt="Mars in colour",
+                alt="",
```
Why remove the alt text?

britneywwc replied:
Thanks for the catch, this must've been left out when rebasing.

jmuzina commented on templates/case-studies/index.html:
```diff
@@ -1,6 +1,5 @@
-{% extends "case-studies/base_case-studies.html" %}
+{% extends "templates/base.html" %}
```
I'm only seeing a header and footer on the demo, no hero content:

<img width="1204" alt="Screenshot 2024-08-16 at 8 57 02 AM" src="https://github.com/user-attachments/assets/d1da7cf6-5620-41a0-9396-a3f102e06fbf">

britneywwc replied:
The example case studies are on https://ubuntu-com-14166.demos.haus/case-studies/example after rebasing a recent merged PR, I've updated the PR description to reflect this. The index page should be an overview of case studies.

jmuzina commented on templates/macros/_macros-quote.jinja:
```diff
@@ -0,0 +1,168 @@
+        <p class="u-text--muted u-no-padding u-no-margin">{{ citation_source_title }}</p>
+      {%- endmacro -%}
+      <div class="p-section--shallow">
+        <hr class="u-hide--large p-rule--muted">
```
```suggestion
        <hr class="u-hide--large is-muted">
```
Minor tweak to fix the missing bottom margin on this rule

jmuzina commented on templates/macros/_macros-quote.jinja:
```diff
@@ -0,0 +1,168 @@
+  {% if has_heading_row %}
+    <!--Optional heading -->
+    <div class="p-section--shallow">
+      <hr class="p-rule--highlight">
```
Don't make this change yet, as it hasn't been fully reviewed, but just flagging this now:

This rule's markup will likely need to be changed to `<hr class="is-highlighted">` and be accompanied by the new `hr.is-highlighted` [variant class](https://github.com/canonical/vanilla-framework/pull/5275#issuecomment-2296935760) in order to create the proper spacing below the highlighted rule. It it awaiting team review.

Lyubo's back from vacation so this shouldn't be waiting too much longer - just wanted to keep you in the loop :)

britneywwc replied:
Thanks for the heads up, I'll keep an eye on the quotes macro PR

jmuzina replied:
The [HR is-highlighted change](https://github.com/canonical/vanilla-framework/pull/5241) was merged, but will be released in [Vanilla 4.16.0](https://github.com/canonical/vanilla-framework/releases/tag/untagged-d8d81ac82b2e58f8dab8) which won't release for a few weeks I think. 

However, this isn't a problem for this PR since you aren't using the quote wrapper with a title (which has a highlighted rule at the top). 

If you find that you need to use the quote wrapper with a title (thus needing the highlighted rule with bottom margin), this can be added to styles temporarily:

```scss
hr.is-highlighted {
  @include vf-highlight-bar($colors--theme--text-default);
  &.is-accent {
    @include vf-highlight-bar($color-accent);
  }
}
```

jmuzina reviewed (commented):
Last couple of comments, then you'll be good on my side once https://github.com/canonical/vanilla-framework/pull/5275 is reviewed and merged and the conversation @pastelcyborg left is resolved

jmuzina commented:
Hi @britneywwc, if you have any thoughts about any of your experience using or creating macros as part of this work, it would be helpful to document them so we can work to improve them. 

If that’s the case, could you record your thoughts in a Google Doc or something similar so we can address any issues? Thanks!!

britneywwc commented on templates/macros/_macros-hero.jinja:
```diff
@@ -0,0 +1,189 @@
+    {%- if layout == '25-75' and has_signpost_image -%}
+      {#- On 25-75 signpost, header top should align with top of the signpost to its left. -#}
+      {% set title_class = "u-no-padding--top" %}
+    {%- elif has_full_width_image -%}
```
Thanks for making the changes @jmuzina! Looking forward to the macros distro epic, will keep a close eye at it :)

britneywwc commented:
Sure thing @jmuzina, I'll try to consolidate it in a google doc and send it over.

lyubomir-popov commented:
- nothing should span 4 columns on desktop, as it is not a multiple of 25%. When there's less than 4, we leave the first 3 columns empty, and start from the fourth.

![image](https://github.com/user-attachments/assets/95882020-a80a-44ff-a9b9-4b73fa5e494e)

- Any `hr` that doesn't span the full width should have the class `is-muted`:

![image](https://github.com/user-attachments/assets/7742ed93-d0ee-45f8-bcfc-4020e91d1771)

- Either the headings should start in the first column, or the lines should start in the fourth so they align:

Before:

![image](https://github.com/user-attachments/assets/4824d9a4-2295-4183-84bd-cb6c0c8b0b11)

After: 

![image](https://github.com/user-attachments/assets/bb377565-c211-488d-b998-5e9b31298aa4)

- can we please make captions muted (p-text--muted I believe):
 
![image](https://github.com/user-attachments/assets/78761aa8-5783-4ed5-90ab-40fa589234f6)

- no need for strip classes on headings, this throws things off the baseline grid, pls remove:

![image](https://github.com/user-attachments/assets/994700c4-2df8-4d73-a897-c58fe9aa7607)

- the quote is a secction in itself, it should havea regular p-section class on the wrapper, to create enough space under itself:
 
![image](https://github.com/user-attachments/assets/c79df749-d362-4440-866c-9a8bd27e4018)

britneywwc commented on templates/macros/_macros-quote.jinja:
```diff
@@ -129,21 +129,23 @@
-          <div class="col-3">
-            {{ _citation_block() }}
+      <hr class="u-hide--medium u-hide--small is-muted">
+      <div class="p-section--shallow">
```
@jmuzina I added a `p-section--shallow` wrapper to the quote_body() per Lyubo's design comment. Lmk if this is okay, otherwise I will create a custom case study quote macro that adds the wrapper to this existing macro.

jmuzina replied:
Hey @britneywwc , I'm checking with Lyubo to see if this is the correct place to put this shallow section, I'll get back to you (and hopefully approve this PR finally) tomorrow.

britneywwc replied:
Thanks! :)

lyubomir-popov commented:
Can we move these 3 cols ot the left, so the body content aligns with the body copy of the article underneath?
![image](https://github.com/user-attachments/assets/13d4a90f-2201-487a-b63c-1a195c2f05a6)

also, are the h2s always going to ber this short?

lyubomir-popov commented:
![image](https://github.com/user-attachments/assets/0d5d9d29-4983-4d41-83e2-0c8e46f7509c)

jmuzina commented:
Pushed a couple of adjustments from code & design review this morning with @pastelcyborg and @lyubomir-popov.

jmuzina reviewed (approved):
Thanks for your work and for being an early macros adopter :)
